packages/vite/src/node/plugins/define.ts

```diff
@@ -122,23 +119,34 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [define, pattern, banner] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern, importMetaEnvVal] = ssr
+        ? ssrPattern
+        : defaultPattern
       if (!pattern) return
 
       // Check if our code needs any replacements before running esbuild
       pattern.lastIndex = 0
       if (!pattern.test(code)) return
 
+      let marker = importMetaEnvMarker
+      while (new RegExp(escapeRegex(marker)).test(code)) {
+        marker += '$'
+      }
+      if (marker !== importMetaEnvMarker && 'import.meta.env' in define) {
+        define['import.meta.env'] = marker
+      }
+
       const result = await replaceDefine(code, id, define, config)
 
       // Replace `import.meta.env.*` with undefined
-      result.code = result.code.replaceAll(importMetaEnvKeyRe, (m) =>
-        'undefined'.padEnd(m.length),
+      result.code = result.code.replaceAll(
+        new RegExp(`${escapeRegex(marker)}\\..+?\\b`, 'g'),
+        (m) => 'undefined'.padEnd(m.length),
```

Comment:
I've not measured, but it feels like creating a regexp and escaping for every replacement is going to be quite perf intensive. Maybe we can use a cache instead?